lib/rubocop/cop/layout/space_inside_array_literal_brackets.rb

```diff
@@ -0,0 +1,241 @@
+      #
+      # @example EnforcedStyle: compact
+      #   # The `compact` style normally requires a space inside
+      #   # array brackets, with the exception that successive left
+      #   # or right brackets are collapsed together in nested arrays.
+      #
+      #   # bad
+      #   array = [ a, [ b, c ] ]
+      #
+      #   # good
+      #   array = [ a, [ b, c ]]
+      class SpaceInsideArrayLiteralBrackets < Cop
+        include SurroundingSpace
+        include ConfigurableEnforcedStyle
+
+        MSG = '%<command>s space inside array brackets.'.freeze
+        EMPTY_MSG = '%<command>s space inside empty array brackets.'.freeze
+
+        def on_array(node)
+          left, right = array_brackets(node)
+          return if right.nil?
+          return empty_offenses(node, left, right) if empty_brkts?(left, right)
+
+          start_ok = next_to_newline?(node, left)
+          end_ok = node.single_line? ? false : end_has_own_line?(right)
+
+          issue_offenses(node, left, right, start_ok, end_ok)
+        end
+
+        private
+
+        def autocorrect(node)
+          left, right = array_brackets(node)
+
+          lambda do |corrector|
+            if empty_brkts?(left, right)
+              empty_corrections(corrector, left, right)
+            elsif style == :no_space
+              no_space_corrector(corrector, left, right)
+            elsif style == :space
+              space_corrector(corrector, left, right)
+            else
+              compact_corrections(corrector, node, left, right)
+            end
+          end
+        end
+
+        def array_brackets(node)
+          [left_array_bracket(node), right_array_bracket(node)]
+        end
+
+        def left_array_bracket(node)
+          tokens(node).find { |token| left_bracket?(token) }
+        end
+
+        def right_array_bracket(node)
+          tokens(node).reverse.find { |token| right_bracket?(token) }
+        end
+
+        def empty_brkts?(left, right)
+          processed_source.tokens.index(left) ==
+            processed_source.tokens.index(right) - 1
+        end
+
+        def empty_offenses(node, left, right)
+          empty_offense(node, 'Use one') if offending_empty_space?(left, right)
+          return unless offending_empty_no_space?(left, right)
+          empty_offense(node, 'Do not use')
+        end
+
+        def empty_offense(node, command)
+          add_offense(node, message: format(EMPTY_MSG, command: command))
+        end
+
+        def offending_empty_space?(left, right)
+          empty_config == 'space' && !space_between?(left, right)
+        end
+
+        def offending_empty_no_space?(left, right)
+          empty_config == 'no_space' && !no_space_between?(left, right)
+        end
+
+        def space_between?(left, right)
+          left.pos.end_pos + 1 == right.pos.begin_pos
+        end
+
+        def no_space_between?(left, right)
+          left.pos.end_pos == right.pos.begin_pos
+        end
```

Comment:
`space_between?` and `no_space_between?` seem like great candidates to move into the `surrounding_space` mixin.  Particularly because there was a `space_between?` method in there. 
But right now these are only used in this class, so here they are.